File: go/bcm_test.go
package bcm_test

import (
	"bytes"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"os"
	"reflect"
	"testing"

	"log"
	"path/filepath"

	bcm "github.com/bitcache-tech/bitcache-message"
)

func GetAllFilesFromDirectory(directory string) ([]string, error) {
	var files []string

	err := filepath.Walk(directory, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			log.Printf("Error accessing path %q: %v\n", path, err)
			return nil
		}
		if !info.IsDir() {
			files = append(files, path)
		}
		return nil
	})

	if err != nil {
		return nil, err
	}

	return files, nil
}

func TestNewBitCacheMessage(t *testing.T) {
	files, err := GetAllFilesFromDirectory("../data/bitcache-messages")
	if err != nil {
		t.Errorf("Error getting files from directory: %v", err)
	}

	for _, file := range files {
		data, err := os.ReadFile(file)
		if err != nil {
			t.Errorf("Error reading file %q: %v", file, err)
			continue
		}
		decoded, err := bcm.NewBitCacheMessage(data)
		if err != nil {
			log.Printf("Error decoding file %q: %v", file, err)
		}

		filename := filepath.Base(file)
		extension := filepath.Ext(file)
		filenameWithoutExt := filename[:len(filename)-len(extension)]

		expectedTxBytes, err := os.ReadFile(fmt.Sprintf("../data/%s.txn", filenameWithoutExt))
		if err != nil {
			t.Errorf("Error reading file %q: %v", file, err)
			continue
		}
		if !bytes.Equal(decoded.Tx.Bytes(), expectedTxBytes) {
			t.Errorf("Decoded BitCacheMessage Tx does not match original for: " + file)

		}
		expectedJSON, err := os.ReadFile(fmt.Sprintf("../data/%s.json", filenameWithoutExt))
		if err != nil {
			t.Errorf("Error reading file %q: %v", file, err)
			continue
		}
		var expectedKUMs []*bcm.KeyUsageMetadata
		json.Unmarshal(expectedJSON, &expectedKUMs)
		if !reflect.DeepEqual(decoded.KUMs, expectedKUMs) {
			t.Errorf("Decoded BitCacheMessage KUM does not match original for: " + file)
		}

	}

}

func Test(t *testing.T) {
	prefix := "TXOKEYS."
	fmt.Println(hex.EncodeToString([]byte(prefix)))
}

File: ts/bcm.test.ts
import { BitCacheMessage, KeyUsageMetadata } from './bcm';
import * as fs from 'fs';
import * as path from 'path';

describe('BitCacheMessage', () => {
    const testDataDir = '../data/bitcache-messages';
    const jsonAndTxDataDir = '../data';

    const removeFileExtension = (filename: string): string => {
        return filename.replace(path.extname(filename), '');
    };

    test('Deserialize binary files and check with JSON and transaction files', () => {
        const binaryFiles = fs.readdirSync(testDataDir)
            .filter(file => file.endsWith('.bin'))
            .map(file => path.join(testDataDir, file));

        for (const file of binaryFiles) {
            const inputBCMBytes = fs.readFileSync(file);
            const decodedBCM = BitCacheMessage.fromBytes(inputBCMBytes);

            const baseFilename = removeFileExtension(path.basename(file));
            const expectedTxBytes = fs.readFileSync(path.join(jsonAndTxDataDir, baseFilename + '.txn'));
            const expectedTxHex = expectedTxBytes.toString('hex');
            const expectedKUMs: KeyUsageMetadata[] = JSON.parse(fs.readFileSync(path.join(jsonAndTxDataDir, baseFilename + '.json'), 'utf-8'));

            expect(decodedBCM.Tx).toEqual(expectedTxHex);
            expect(decodedBCM.KUMs).toEqual(expectedKUMs);
        }
    });

    test('Encode JSON and transaction files to binary and compare with binary files', () => {
        const jsonFiles = fs.readdirSync(jsonAndTxDataDir)
            .filter(file => file.endsWith('.json'))
            .map(file => path.join(jsonAndTxDataDir, file));

        for (const file of jsonFiles) {
            const baseFilename = removeFileExtension(path.basename(file));
            const inputTxBytes = fs.readFileSync(path.join(jsonAndTxDataDir, baseFilename + '.txn'));
            const inputKUMs: KeyUsageMetadata[] = JSON.parse(fs.readFileSync(file, 'utf-8'));

            const bcm = new BitCacheMessage(inputTxBytes.toString('hex'), inputKUMs);
            const encodedData = bcm.toBytes();

            const expectedBCMBytes = fs.readFileSync(path.join(testDataDir, baseFilename + '.bin'));
            expect(encodedData).toEqual(expectedBCMBytes);
        }
    });
});
